### PR TITLE
LoongArch: support relaxation of pcalau12i/ld.d to pcalau12i/addi.d or pcaddi

### PR DESCRIPTION
**Pull request by ywgrit**
If we get symval by accessing to local got, then we can omit one memory loading.
The first phase: pcalau12i/ld.d to pcala12i/addi.d
The second phase: pcalau12i/addi.d to pcaddi

### CODE REVIEW

**Comment by ywgrit** [comment]
@MQ-mengqing

File: elf/arch-loongarch.cc [hunk 1]
```diff
@@ -816,6 +878,39 @@ void shrink_section(Context<E> &ctx, InputSection<E> &isec, bool use_rvc) {
...
+          ctx.arg.relax &&
+          i + 3 < rels.size() &&
+          rels[i + 2].r_type == R_LARCH_GOT_PC_LO12 &&
+          rels[i + 2].r_offset == rels[i].r_offset + 4 &&
+          rels[i + 3].r_type == R_LARCH_RELAX) {
+          u32 insn1 = *(ul32 *)(isec.contents.data() + rels[i].r_offset);
+          u32 insn2 = *(ul32 *)(isec.contents.data() + rels[i].r_offset + 4);
+
+          // relax pcalau12i/ld.d to pcalau12i/addi.d
+          if (get_rd(insn1) != get_rd(insn2))
```
Comment (MQ-mengqing) [inline elf/arch-loongarch.cc:905]: For more reliable, check rj is also necessary.

File: elf/arch-loongarch.cc [hunk 2]
```diff
@@ -816,6 +878,39 @@ void shrink_section(Context<E> &ctx, InputSection<E> &isec, bool use_rvc) {
...
+            continue;
+
+          i64 dist = compute_distance(ctx, sym, isec, r);
+          // the second phase: relax pcalau12i/addi.d to pcaddi
+          if (dist % 4 == 0 && -(1 << 21) < dist && dist < (1 << 21))
```
Comment (MQ-mengqing) [inline elf/arch-loongarch.cc:910]: There should be an eage case "-(1<<21) == dist".

File: elf/arch-loongarch.cc [hunk 3]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
+            !sym.is_absolute() &&
+            !sym.is_ifunc() &&
+            ctx.arg.relax &&
+            rels[i+1].r_type == R_LARCH_RELAX &&
+            rels[i+3].r_type == R_LARCH_RELAX &&
+            rels[i+2].r_type == R_LARCH_GOT_PC_LO12 &&
+            rels[i+2].r_offset == rel.r_offset + 4) {
+            u32 insn1 = *(ul32 *)(contents.data() + rel.r_offset);
+            u32 insn2 = *(ul32 *)(contents.data() + rels[i+2].r_offset);
+            u32 rd = get_rd(insn1);
```
Comment (MQ-mengqing) [inline elf/arch-loongarch.cc:418]: I think there should be a range check for pcalau12i.

File: elf/arch-loongarch.cc [hunk 4]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
+
+            if (rd == get_rd(insn2)) {
+              // pcalau12i/ld.d has been relaxed to pcalau12i/addi.d
+              // reloc the pcalau12i as R_LARCH_PLACA_HI20
+              write_j20(loc, hi20(S + A, P));
+              break;
```
Comment (rui314) [inline elf/arch-loongarch.cc:424]: You should directly write to `loc + 4` and then `i += 3` here, instead of doing the same check in `case R_LARCH_GOT_PC_LO12`.

File: elf/arch-loongarch.cc [hunk 5]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
-      break;
+      switch (removed_bytes) {
+        // pcalau12i/ld.d has been relaxed to pcaddi, the first insn has been removed.
+        case 4:
+          break;
+        case 0:
+          if (i + 3 < rels.size() &&
+            sym.is_local(ctx) &&
+            !sym.is_absolute() &&
+            !sym.is_ifunc() &&
```
Comment (rui314) [inline elf/arch-loongarch.cc:410]: It looks like you want to do the same as we do for `R_RISCV_GOT_HI20`. Did you check that in arch-riscv.cc? We have `sym.is_pcrel_linktime_const(ctx)` to check if the symbol's PC-relative address is link-time constant.

File: elf/arch-loongarch.cc [hunk 6]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
+            ctx.arg.relax &&
+            rels[i+1].r_type == R_LARCH_RELAX &&
```
Comment (rui314) [inline elf/arch-loongarch.cc:412]: You want to pay more attention to the details -- add spaces around `+`. This is how we format it in other places.

File: elf/arch-loongarch.cc [hunk 7]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
+            rels[i+3].r_type == R_LARCH_RELAX &&
+            rels[i+2].r_type == R_LARCH_GOT_PC_LO12 &&
+            rels[i+2].r_offset == rel.r_offset + 4) {
+            u32 insn1 = *(ul32 *)(contents.data() + rel.r_offset);
```
Comment (rui314) [inline elf/arch-loongarch.cc:416]: and also indentation.

File: test/elf/loongarch64_relax-got.sh [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  ld.w $a0, $a0, 0
+  ret
+get_sym4:
+  la.global $a0, sym4
+  ld.w $a0, $a0, 0
+  ret
+get_sym5:
+  la.global $a0, sym5
+  ld.w $a0, $a0, 0
+  ret
```
Comment (rui314) [inline test/elf/loongarch64_relax-got.sh:25]: I don't think you need five symbols for this test, as they are testing exactly the same relaxation.

File: elf/arch-loongarch.cc [hunk 9]
```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
...
+
+          if (rd == get_rd(insn2)) {
+            // pcalau12i/ld.d has been relaxed to pcalau12i/addi.d
+            // rewrite the ld.d with addi.d
+            *(ul32 *)loc = 0x02c00000 | rd | (rd << 5);
+            write_k12(loc, S + A);
+            break;
+          }
+        }
+
```
Comment (rui314) [inline elf/arch-loongarch.cc:396]: I think you can revert this change.

**Comment by ywgrit** [comment]
Do not care about the last two commits now, there are still some issues, I'll commit the new code once I've tested it out.

**Comment by rui314** [comment]
I'll fix a few minor issues after merging this change. Next time, please pay more attention to indentation. Thank you!

**Comment by ywgrit** [comment]
> I'll fix a few minor issues after merging this change. Next time, please pay more attention to indentation. Thank you!

Yeah, Sorry for that.

**Comment by ywgrit** [comment]
> Thank you for doing this anyway. Did you have anything other than this, or is this the last one?


TLS LD/GD relaxation is the next plan.

**Comment by rui314** [comment]
I actually don't think you can relax TLS GD/LD on LoongArch, because calls to `__tls_get_addr` is not annotated by any relocation, so you cannot eliminate that function call. You might still be able to relax a R_LARCH_TLS_GD_PC_HI20 and R_LARCH_GOT_PC_LO12 pair into a single instruction, but that optimization is marginal and ineffective. `__tls_get_addr` is expensive anyways, so I wouldn't try to do that relaxation. What were you planning to do?

**Comment by ywgrit** [comment]
> I actually don't think you can relax TLS GD/LD on LoongArch, because calls to `__tls_get_addr` are not annotated by any relocation, so you cannot eliminate that function call. You might still be able to relax a R_LARCH_TLS_GD_PC_HI20 and R_LARCH_GOT_PC_LO12 pair into a single instruction, but that optimization is marginal and probably meaningless. `__tls_get_addr` is expensive anyways, so I wouldn't try to do that relaxation. What were you planning to do?

You're right, actually tls_desc is currently supported by loongarch, though it hasn't been introduced on mold yet. Considering the negligible performance gain from saving an instruction, it seems there is no need to apply relaxation on tls.

**Comment by rui314** [comment]
Then maybe you want to add TLSDESC support?

**Comment by ywgrit** [comment]
> Then maybe you want to add TLSDESC support?

Yeah, I want have a try.
